src/software/geom/algorithms/rasterize.cpp

```diff
@@ -0,0 +1,245 @@
+
+    // Draw points resolution_size away (arclength) from each other over the circumference
+    double delta_theta     = resolution_size / radius;
+    int num_theta_sections = (int)std::ceil(M_PI / delta_theta);
+    for (int section = 0; section < num_theta_sections; section++)
+    {
+        double theta    = section * delta_theta;
+        double x_offset = std::cos(theta) * radius;
+        double y_offset = std::sin(theta) * radius;
+
+        covered_points.emplace_back(Point(origin.x() + x_offset, origin.y() + y_offset));
+        covered_points.emplace_back(Point(origin.x() + x_offset, origin.y() - y_offset));
+    }
+
+    return covered_points;
+}
+
+std::vector<Point> rasterize(const Rectangle& rectangle, const double resolution_size)
+{
+    // Loop through the length and the width of the rectangle and add points
+    // resolution_size away from each other. Also making sure that the edges are also
+    // filled with points if the dimensions of the rectangle are not divisible by
+    // resolution_size.
+    std::vector<Point> covered_points;
+
+    int num_pixels_x = (int)std::ceil(rectangle.xLength() / resolution_size);
+    int num_pixels_y = (int)std::ceil(rectangle.yLength() / resolution_size);
+
+    for (int x_pixel = 0; x_pixel <= num_pixels_x; x_pixel++)
+    {
+        // x and y offset from the top left corner of the rectangle
+        double x_offset;
+
+        // Adjust the last x and y pixels to be on the edge of the rectangle to make sure
+        // that the points cover the entire rectangle without going outside.
+        if (x_pixel == num_pixels_x)
+        {
+            x_offset = rectangle.xLength();
+        }
+        else
+        {
+            x_offset = x_pixel * resolution_size;
+        }
+
+        for (int y_pixel = 0; y_pixel <= num_pixels_y; y_pixel++)
+        {
+            double y_offset;
+            if (y_pixel == num_pixels_y)
+            {
+                y_offset = rectangle.yLength();
+            }
+            else
+            {
+                y_offset = y_pixel * resolution_size;
+            }
+
+            covered_points.emplace_back(
+                Point(rectangle.xMin() + x_offset, rectangle.yMin() + y_offset));
+        }
+    }
+
+    return covered_points;
+}
+
+std::vector<Point> rasterize(const Polygon& polygon, const double resolution_size)
+{
+    // This rasterize function uses two steps to fully cover a polygon.
+    // 1. Draw a bounding box around the polygon and loop through it using a similar
+    // approach as rasterize for polygon and only add the points that are contained by the
+    // polygon.
+    //                                     AND
+    // 2. To ensure that the edges are covered, draw a line between every two consecutive
+    // vertices and draw points on the line, resolution_size or less away from each other.
+
+    // Used to avoid points being outside of polygon due to floating point errors
+    const double EPSILON         = 0.0001;
```

Comment:
is it possible to use FIXED_EPSILON from geom_constants.h

Reply:
I switched to using FIXED_EPSILON instead, and all of the rasterize_tests continue passing, showing that no point is added outside of the circle due to floating point errors.